apps/emqx/src/emqx_broker.erl

```diff
@@ -160,15 +160,20 @@ do_subscribe(Topic, SubPid, SubOpts) ->
 do_subscribe2(Topic, SubPid, SubOpts) when is_binary(Topic) ->
     %% FIXME: subscribe shard bug
     %% https://emqx.atlassian.net/browse/EMQX-10214
-    case emqx_broker_helper:get_sub_shard(SubPid, Topic) of
+    {Shard, Seq} = emqx_broker_helper:get_sub_shard_and_seq(SubPid, Topic),
+    IsFirst = Seq =:= 1,
+    case Shard of
         0 ->
             true = ets:insert(?SUBSCRIBER, {Topic, SubPid}),
             true = ets:insert(?SUBOPTION, {{Topic, SubPid}, SubOpts}),
-            call(pick(Topic), {subscribe, Topic});
+            case IsFirst of
+                true -> call(pick(Topic), {subscribe, Topic});
```

Comment:
This is probably too optimistic: if the first subscriber failed to register the route, the next one won't even try it.. :face_with_head_bandage: 
Need a more robust solution.

Reply:
Maybe should keep `emqx_router:has_route/2` check but skip it for the first subscriber, so that the first one always adds a route (even if the previous stale route is still present). 
Or just serialize writes to `?SUBSCRIBER` tab by means of `emqx_broker` pool...

Reply:
Another possible solution is to implement new Mria API that will synchronously update local replicant state and avoid core->replicant replication for the node that initiated a write op.

Reply:
> Or just serialize writes to `?SUBSCRIBER` tab by means of emqx_broker pool...

I'm also starting to think about going into this direction.